File: 2021_7/exist.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        h, w = len(board), len(board[0])
        visit = [[0 for _ in range(w)] for _ in range(h)] # whether visited in this term
        def inside(i, j):
            if (i >=0 and i < h) and (j >=0 and j < w):
                return True
            else:
                return False

        def recursive_exist(i, j, temp):
            if temp == word:
                return True
            if len(temp) == len(word):
                return False
            cur = len(temp)
            if inside(i,j):
                if word[cur] == board[i][j] and not visit[i][j]: # current char equals and not visited, recursive from four directins
                    visit[i][j] = 1
                    temp += word[cur]
                    #up 
                    if recursive_exist(i-1, j, temp):
                        return True
                    if recursive_exist(i+1, j, temp):
                        return True
                    if recursive_exist(i, j-1, temp):
                        return True
                    if recursive_exist(i, j+1, temp):
                        return True
                    temp = temp[:-1]
                    visit[i][j] = 0
                    return False
            return False


        for row in range(h):
            for col in range(w):
                temp = ""
                if recursive_exist(row, col, temp):
                    return True
        
        return False
    
board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]]
# word = "ABCCED"
word = "ABCB"
s = Solution()
print(s.exist(board, word))

File: leetcode_hot100/exist.cc
#include "head.h"
class Solution
{
public:
    bool exist(vector<vector<char>> &board, string word)
    {
        int l_x = board.size(), l_y = board[0].size();
        for (int x = 0; x < l_x; ++x)
        {
            for (int y = 0; y < l_y; ++y)
            {
                vector<vector<bool>> visit(l_x, vector<bool>(l_y, false));
                if (dfs_search(board, x, y, word, 0, visit))
                    return true;
            }
        }
        return false;
    }

    bool dfs_search(vector<vector<char>> &board, int x,
                    int y, string &word, int idx, vector<vector<bool>> &visit)
    {
        if (idx == word.size())
        {
            return true;
        }
        if (x < 0 || x >= board.size() || y < 0 || y > board[0].size())
        {
            return false;
        }
        if (visit[x][y] || board[x][y] != word[idx])
        {
            return false;
        }
        /* board[x][y] == word[idx] */
        visit[x][y] = true;
        bool m_exist = dfs_search(board, x + 1, y, word, idx + 1, visit) ||
                       dfs_search(board, x - 1, y, word, idx + 1, visit) ||
                       dfs_search(board, x, y + 1, word, idx + 1, visit) ||
                       dfs_search(board, x, y - 1, word, idx + 1, visit);
        visit[x][y] = false;
        return m_exist;
    }
};